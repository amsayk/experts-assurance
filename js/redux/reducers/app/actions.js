import {
  RESIZE,
  CONNECTION_STATE_CHANGE,
  APP_STATE_CHANGE,
  READY,
  TOGGLE_ALERTS,

  ADD_DOC,
} from './constants';

export function resize() {
  return {
    type: RESIZE,
  };
}

export function connectionStateChange(isConnected) {
  return {
    type: CONNECTION_STATE_CHANGE,
    isConnected,
  };
}

export function appStateChange(currentState) {
  return {
    type: APP_STATE_CHANGE,
    currentState,
  };
}

export function ready() {
  return {
    type: READY,
  };
}

export function toggleAlerts() {
  return {
    type: TOGGLE_ALERTS,
  };
}

export function startAddingDoc() {
  return {
    type      : ADD_DOC,
    addingDoc : true,
  };
}

export function finishAddingDoc() {
  return {
    type      : ADD_DOC,
    addingDoc : false,
  };
}

export function saveNewDoc(data) {
  return async (dispatch, _, { client }) => {

  };
}

