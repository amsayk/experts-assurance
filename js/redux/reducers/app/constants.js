import keyOf from 'keyOf';

export const READY = keyOf({READY: null});
export const RESIZE = keyOf({RESIZE: null});
export const CONNECTION_STATE_CHANGE = keyOf({CONNECTION_STATE_CHANGE: null});

