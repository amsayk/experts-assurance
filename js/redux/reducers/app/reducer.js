import {
  READY,
  RESIZE,
  CONNECTION_STATE_CHANGE,
} from './constants';

import { isServer } from 'vars';

import Immutable from 'immutable';

const MEDIA_QUERY = '(min-width: 992px)';

const initialState = Immutable.fromJS({
  displayMatches : isServer || matchMedia(MEDIA_QUERY).matches,
  onLine         : true,
  isReady        : false,
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case RESIZE:
      return state.merge({
        displayMatches: matchMedia(MEDIA_QUERY).matches,
      });
    case CONNECTION_STATE_CHANGE:
      return state.merge({
        onLine: action.isConnected,
      });
    case READY:
      return state.merge({
        isReady: true,
      });
    default:
      return state;
  }
}

