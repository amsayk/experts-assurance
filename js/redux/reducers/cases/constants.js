import keyOf from 'keyOf';

import {
  ACTION as SORT_ACTION,
} from 'redux/reducers/sorting/constants';

import {
  ACTION as SELECTION_ACTION,
} from 'redux/reducers/selection/constants';

// ------------------------------------
// Constants
// ------------------------------------

export const ON_SEARCH = keyOf({CASES_ON_SEARCH: null});
export const ON_STATE = keyOf({CASES_ON_STATE: null});
export const ON_CLIENT = keyOf({CASES_ON_CLIENT: null});
export const ON_INSURER = keyOf({CASES_ON_INSURER: null});

export const SORT = keyOf({[`${SORT_ACTION}/cases`]: null});
export const SELECTION = keyOf({[`${SELECTION_ACTION}/cases`]: null});

