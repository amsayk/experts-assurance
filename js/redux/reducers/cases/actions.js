import {
  ON_SEARCH,
  ON_STATE,
  ON_CLIENT,
  ON_MANAGER,
  ON_AGENT,
} from './constants';

// Selection
import { set, add, toggle } from 'redux/reducers/selection/actions';
export const toggleSelection = toggle('cases');
export const setSelection    = set('cases');
export const addToSelection  = add('cases');

// Sorting
import { sort, sortKey, sortDirection } from 'redux/reducers/sorting/actions';
export const sortCases            = sort('cases');
export const sortCasesByKey       = sortKey('cases');
export const sortCasesByDirection = sortDirection('cases');

export function search(queryString) {
  return {
    type : ON_SEARCH,
    queryString,
  };
}

export function onClient(id) {
  return {
    type : ON_CLIENT,
    id,
  };
}

export function onManager(id) {
  return {
    type : ON_MANAGER,
    id,
  };
}

export function onAgent(id) {
  return {
    type : ON_AGENT,
    id,
  };
}

export function onState(state) {
  return {
    type : ON_STATE,
    state,
  };
}

