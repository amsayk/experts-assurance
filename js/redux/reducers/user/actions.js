import Parse from 'parse';
import { USER_LOGGED_IN, USER_LOGGED_OUT } from './constants';

import debug from 'log';

const log = debug('app:client:auth');

export function login(payload) {
  return {
    type: USER_LOGGED_IN,
    payload,
  };
}

export function logOut(manual = true) {
  return async (dispatch, _, { client }) => {
    try {
      await Parse.User.logOut();
    } catch (e) {
      switch (e.code) {
        case Parse.Error.INVALID_SESSION_TOKEN:
          log.error('INVALID_SESSION_TOKEN: clearing storage.');
          return Parse.CoreManager.getStorageController().clear();
        default:
          log.error(e);
      }
    } finally {
      dispatch({ type: USER_LOGGED_OUT, manual });
      // client.resetStore();
    }
  };
}

