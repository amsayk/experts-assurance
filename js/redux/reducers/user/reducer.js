import getCurrentUser from 'getCurrentUser';

import { fromJS } from 'immutable';

import { USER_LOGGED_IN, USER_LOGGED_OUT } from './constants';

import { INIT } from 'vars';

function maybeUser() {
  const user = getCurrentUser();
  return user ? {
    id: user.id,
    email: user.get('email'),
  } : {};
}

const initialState = fromJS({});

export default function userReducer(state = initialState, { type, payload }) {
  if (type === USER_LOGGED_IN) {
    return fromJS(payload);
  }
  if (type === USER_LOGGED_OUT) {
    return fromJS({});
  }
  if (type === INIT) {
    return fromJS(maybeUser());
  }
  return state;
}

