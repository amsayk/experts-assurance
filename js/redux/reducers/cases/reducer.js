import cookie from 'react-cookie';

import {
  SELECTION,
  SORT,

  ON_SEARCH,
  ON_STATE,
  ON_CLIENT,
  ON_MANAGER,
  ON_AGENT,
} from './constants';

import { USER_LOGGED_OUT, USER_LOGGED_IN } from 'redux/reducers/user/constants';

import { INIT } from 'vars';

// selection
import selectionReducer, {
  initialState as selectionInitialState,
} from 'redux/reducers/selection/reducer';

// sort
import sortReducer, {
  initialState as sortInitialState,
} from 'redux/reducers/sorting/reducer';

import { Record } from 'immutable';

const initialSortConfig = sortInitialState.merge({ key : 'dateMission' });

export class CasesState extends Record({
  selection    : selectionInitialState,
  sortConfig   : initialSortConfig,
  queryString  : '',
  state        : null,
  client       : null,
  manager      : null,
  agent        : null,
}) {}

const initialState = new CasesState();

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case SELECTION: {
      return state.merge({
        selection: selectionReducer(state.selection, action.payload),
      });
    }
    case SORT: {
      return state.merge({
        sortConfig: sortReducer(state.sortConfig, action.payload),
      });
    }
    case ON_SEARCH: {
      return state.merge({
        queryString: action.queryString,
      });
    }
    case ON_STATE: {
      return state.merge({
        state: action.state,
      });
    }
    case ON_CLIENT: {
      return state.merge({
        client: action.id,
      });
    }
    case ON_MANAGER: {
      return state.merge({
        manager: action.id,
      });
    }
    case ON_AGENT: {
      return state.merge({
        agent: action.id,
      });
    }

      case USER_LOGGED_IN:
    case INIT: {
      return state.merge({
        sortConfig : initialSortConfig.merge(cookie.load('cases.sortConfig', /* doNotParse = */false)),
      });
    }
    case USER_LOGGED_OUT:
      return action.manual ? initialState : state;
    default:
      return state;
  }
}

