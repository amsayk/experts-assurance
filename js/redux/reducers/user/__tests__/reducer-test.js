jest.mock('common/getCurrentUser');

import { login } from '../actions';
import { USER_LOGGED_OUT } from '../constants';
import reducer from '../reducer';

import { fromJS } from 'immutable';

import { INIT } from 'vars';

import getCurrentUser from 'getCurrentUser';

const randomActionType = '@@UNKNOWN_ACTION';

describe('user reducer', () => {
  it('should return initialState on unhandled action', () => {
    expect(reducer(undefined, { type: randomActionType }))
      .toEqual(fromJS({}));

    expect(reducer(undefined, { type: randomActionType }))
      .toMatchSnapshot();
  });

  it('should mentain its state on unhandled action', () => {
    expect(reducer(fromJS({ id: 'myId' }), { type: randomActionType }))
      .toEqual(fromJS({
        id: 'myId',
      }));

    expect(reducer(fromJS({ id: 'myId' }), { type: randomActionType }))
      .toMatchSnapshot();
  });

  it('should handle `INIT`', () => {
    getCurrentUser._setCurrentUser({ id: '1' });
    expect(reducer(undefined, { type: INIT }))
      .toEqual(fromJS({
        id: '1',
      }));

    expect(reducer(undefined, { type: INIT }))
      .toMatchSnapshot();

    getCurrentUser._setCurrentUser(null);
    expect(reducer(undefined, { type: INIT }))
      .toEqual(fromJS({}));

    expect(reducer(undefined, { type: INIT }))
      .toMatchSnapshot();
  });

  it('should handle login', () => {
    const user = { id: 'myLoginId' };
    expect(reducer(undefined, login(user)))
      .toEqual(fromJS(user));

    expect(reducer(undefined, login(user)))
      .toMatchSnapshot();
  });

  it('should handle logOut', () => {
    expect(reducer(undefined, { type: USER_LOGGED_OUT }))
      .toEqual(fromJS({}));

    expect(reducer(undefined, { type: USER_LOGGED_OUT }))
      .toMatchSnapshot();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });
});

