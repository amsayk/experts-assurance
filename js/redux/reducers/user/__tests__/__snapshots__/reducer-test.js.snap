exports[`user reducer should handle \`INIT\` 1`] = `
Object {
  "email": "email",
  "id": "1",
}
`;

exports[`user reducer should handle \`INIT\` 2`] = `Object {}`;

exports[`user reducer should handle logOut 1`] = `Object {}`;

exports[`user reducer should handle login 1`] = `
Object {
  "id": "myLoginId",
}
`;

exports[`user reducer should mentain its state on unhandled action 1`] = `
Object {
  "id": "myId",
}
`;

exports[`user reducer should return initialState on unhandled action 1`] = `Object {}`;
